controller/pod/promotionhandler.go

```diff
@@ -0,0 +1,165 @@
+	"fmt"
+	"strings"
+	"time"
+
+	crv1 "github.com/crunchydata/postgres-operator/apis/cr/v1"
+	"github.com/crunchydata/postgres-operator/config"
+	"github.com/crunchydata/postgres-operator/kubeapi"
+	"github.com/crunchydata/postgres-operator/operator/backrest"
+	clusteroperator "github.com/crunchydata/postgres-operator/operator/cluster"
+	log "github.com/sirupsen/logrus"
+	apiv1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+)
+
+// handlePostgresPodPromotion is responsible for handling updates to PG pods the occur as a result
+// of a failover.  Specifically, this handler is triggered when a replica has been promoted, and
+// it now has either the "promoted" or "master" role label.
+func (c *Controller) handlePostgresPodPromotion(newPod *apiv1.Pod, clusterName string) error {
+
+	if err := cleanAndCreatePostFailoverBackup(c.PodClient, c.PodClientset,
+		clusterName, newPod.Namespace); err != nil {
+		log.Error(err)
+		return err
+	}
+
+	return nil
+}
+
+// handleStandbyPodPromotion is responsible for handling updates to PG pods the occur as a result
+// of disabling standby mode.  Specifically, this handler is triggered when a standy leader
+// is turned into a regular leader.
+func (c *Controller) handleStandbyPromotion(newPod *apiv1.Pod, cluster crv1.Pgcluster) error {
+
+	clusterName := cluster.Name
+	namespace := cluster.Namespace
+
+	if err := waitForStandbyPromotion(c.PodConfig, c.PodClientset, *newPod, cluster); err != nil {
+		return err
+	}
+
+	// rotate the pgBouncer passwords if pgbouncer is enabled within the cluster
+	if cluster.Labels[config.LABEL_PGBOUNCER] == "true" {
+		parameters := map[string]string{
+			config.LABEL_PGBOUNCER_ROTATE_PASSWORD: "true",
+			config.LABEL_PGBOUNCER_TASK_CLUSTER:    cluster.Name,
+		}
+		if err := clusteroperator.CreatePgTaskforUpdatepgBouncer(c.PodClient, &cluster,
+			"", parameters); err != nil {
+			log.Error(err)
+			return err
+		}
+	}
+
+	if err := cleanAndCreatePostFailoverBackup(c.PodClient, c.PodClientset, clusterName,
+		namespace); err != nil {
+		log.Error(err)
+		return err
+	}
+
+	return nil
+}
+
+// waitForStandbyPromotion waits for standby mode to be disabled for a specific cluster and has
+// been promoted.  This is done by verifying that recovery is no longer enabled in the database,
+// while also ensuring there are not any pending restarts for the database.
+// done by confirming
+func waitForStandbyPromotion(restConfig *rest.Config, clientset *kubernetes.Clientset, newPod apiv1.Pod,
+	cluster crv1.Pgcluster) error {
+
+	// Checks to see if the DB is in recovery
+	isInRecoveryCMD := make([]string, 4)
+	isInRecoveryCMD[0] = "psql"
+	isInRecoveryCMD[1] = "-t"
+	isInRecoveryCMD[2] = "-c"
+	isInRecoveryCMD[3] = "select pg_is_in_recovery()"
+
+	// Pulls back details about the master
+	leaderStatusCMD := make([]string, 2)
+	leaderStatusCMD[0] = "curl"
+	leaderStatusCMD[1] = fmt.Sprintf("localhost:%s/master", config.DEFAULT_PATRONI_PORT)
```

Comment:
we could put both of this at the top of the file as "constants" i.e. `var` so we don't have to execute this every time :-) the pgbouncer code I believe has some examples

Reply:
like

```go
var isInRecoveryCommand := []string{"psql", "-t", "-c", "'SELECT pg_is_in_recovery();'"}
```

Reply:
Moved the to vars at the top of the file.